junit-jupiter-api/src/main/java/org/junit/jupiter/api/DisplayNameGeneration.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2015-2018 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.api;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+import static org.junit.jupiter.api.DisplayNameGenerator.parameterTypesAsString;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.lang.reflect.Method;
+
+import org.apiguardian.api.API;
+import org.junit.platform.commons.util.Preconditions;
+
+/**
+ * {@code @DisplayNameGeneration} is used to declare a custom display name
+ * generator for the annotated test class.
+ *
+ * <p>Users may select either a pre-defined {@link Style} or supply a
+ * custom {@link DisplayNameGenerator} implementation.
+ *
+ * @since 5.4
+ * @see DisplayName
+ * @see DisplayNameGenerator
+ */
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Inherited
+@API(status = EXPERIMENTAL, since = "5.4")
+public @interface DisplayNameGeneration {
+
+	/**
+	 * Custom display name generator.
+	 *
+	 * @return custom display name generator implementation or {@link DisplayNameGenerator}
+	 *         to use the {@link Style} provided by the {@link #value()} property
+	 */
+	Class<? extends DisplayNameGenerator> generator() default DisplayNameGenerator.class;
+
+	/**
+	 * The pre-defined style to use.
+	 *
+	 * @return the style to use, can be overridden by a custom display
+	 * name {@link #generator()} implementation
+	 */
+	Style value() default Style.DEFAULT;
```

Comment:
Do we need the two separate properties or could we define `DEFAULT` and `UNDERSCORE` as `Class<? extends DisplayNameGenerator>` constants in `DisplayNameGenerator`?

Reply:
That's also an option, but then it might make sense to get rid of the enum and rework the current _delegation_ for reused features.

Actually, I haven't put too much thought into it, but would that change require that we get rid of the enum altogether?

Reply:
Rationale: I'm concerned about instantiating the enum type via reflection, b/c I assume that's not permitted.

Reply:
Moving types as Marc suggested for reviewing it in a dedicated commit - may revert it, if we don't like it.

Reply:
See diff in https://github.com/junit-team/junit5/pull/1588/commits/6e7f2a3325c31cbb85882f7c097dadd14a9e4341 -- only drawback is that only the default generator is held as an internal singleton instance. All other pre-defined generators (that is only `ReplaceUnderscores` for now) need to be instantiated at runtime.

Reply:
We could always introduce a static cache to overcome that drawback (map of instances keyed by FQCN).